packages/aws-cdk/lib/api/nested-stack-helpers.ts

```diff
@@ -0,0 +1,142 @@
+    const deployedNestedStackResource = deployedParentTemplate.Resources[nestedStackLogicalId] ?? {};
+    deployedParentTemplate.Resources[nestedStackLogicalId] = deployedNestedStackResource;
+    deployedNestedStackResource.Type = deployedNestedStackResource.Type ?? 'AWS::CloudFormation::Stack';
+    deployedNestedStackResource.Properties = deployedNestedStackResource.Properties ?? {};
+    deployedNestedStackResource.Properties.NestedTemplate = nestedStackTemplates.deployedTemplate;
+
+    nestedStackNames[nestedStackLogicalId] = {
+      stackName: nestedStackTemplates.deployedStackName,
+      children: await addNestedTemplatesToGeneratedAndDeployedStacks(
+        rootStackArtifact,
+        sdk,
+        nestedStackTemplates,
+      ),
+    };
+  }
+
+  return nestedStackNames;
+}
+
+async function getNestedStackTemplates(
+  rootStackArtifact: cxapi.CloudFormationStackArtifact, nestedTemplateAssetPath: string, nestedStackLogicalId: string,
+  listStackResources: ListStackResources | undefined, sdk: ISDK,
+): Promise<StackTemplates> {
+  const nestedTemplatePath = path.join(rootStackArtifact.assembly.directory, nestedTemplateAssetPath);
+
+  // CFN generates the nested stack name in the form `ParentStackName-NestedStackLogicalID-SomeHashWeCan'tCompute,
+  // the arn is of the form: arn:aws:cloudformation:region:123456789012:stack/NestedStackName/AnotherHashWeDon'tNeed
+  // so we get the ARN and manually extract the name.
+  const nestedStackArn = await getNestedStackArn(nestedStackLogicalId, listStackResources);
+  const deployedStackName = nestedStackArn?.slice(nestedStackArn.indexOf('/') + 1, nestedStackArn.lastIndexOf('/'));
+
+  return {
+    generatedTemplate: JSON.parse(fs.readFileSync(nestedTemplatePath, 'utf-8')),
+    deployedTemplate: deployedStackName
+      ? await loadCurrentStackTemplate(deployedStackName, sdk)
+      : {},
+    deployedStackName,
+  };
+}
+
+async function getNestedStackArn(
+  nestedStackLogicalId: string, listStackResources?: ListStackResources,
+): Promise<string | undefined> {
+  try {
+    const stackResources = await listStackResources?.listStackResources();
+    return stackResources?.find(sr => sr.LogicalResourceId === nestedStackLogicalId)?.PhysicalResourceId;
+  } catch (e) {
+    if (e.message.startsWith('Stack with id ') && e.message.endsWith(' does not exist')) {
+      return;
+    }
+    throw e;
+  }
+}
+
+function isCdkManagedNestedStack(stackResource: any): stackResource is NestedStackResource {
+  return stackResource.Type === 'AWS::CloudFormation::Stack' && stackResource.Metadata && stackResource.Metadata['aws:asset:path'];
+}
+
+export interface TemplateWithNestedStackNames {
+  readonly deployedTemplate: Template;
+  readonly nestedStackNames: { [nestedStackLogicalId: string]: NestedStackNames };
+}
+
+export interface NestedStackNames {
+  readonly stackName: string | undefined;
+  readonly children: { [logicalId: string]: NestedStackNames };
```

Comment:
`children` seems too generic of a name here. `childNestedStackNames` perhaps, or something similar?